### PR TITLE
Items exploding while thrown admin option (or on arcane tampered floors)

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[content]

## What this does
makes any thrown item with this setting on delete itself in a harmless explosion (or if on specific arcane tampered floor tiles, but not if the wizard who tampered it throws it on that tile)

## Why it's good
looks funny, should be rare enough to not cause problems (maybe i can remove the arcane tampered floors bit if that's skubby)

## Changelog
:cl:
 * rscadd: Gods can now cause any thrown item to explode on impact. (or wizards, but only on specific floor tiles and not if thrown at by the wizard who cursed the tile)

### CODE REVIEW

**Comment by pedr69** [comment]
>tamper a turf
>throw nuke disk
>????
>shit bricks

cerebral PR

**Comment by Eneocho** [comment]
The nuke disk is snowflaked to respawn somewhere on the station if destroyed.

**Comment by Exxion** [comment]
Wouldn't this make way more sense to check in `throw_impact()`?

**Comment by SECBATON-GRIFFON** [comment]
> Wouldn't this make way more sense to check in `throw_impact()`?

i'm trying to make it do something on whenever throwing is stopped, i don't think floors call it do they?

**Comment by Exxion** [comment]
They do and there's even a comment explaining this

**Comment by Exxion** [comment]
You didn't read the comment I mentioned. Without checking if `impacted_atom` is a turf, it'll get called once for every atom on the turf the object stops on.

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -213,6 +213,18 @@
...
 	qdel(src)
 
+var/global/objects_thrown_when_explode = FALSE
+
+/obj/item/throw_impact(atom/impacted_atom, speed, mob/user)
+	..()
+	if(isturf(loc))
+		var/turf/T = loc
+		if(objects_thrown_when_explode || T.arcanetampered)
+			if(T)
```
Comment (Exxion) [inline code/game/objects/items.dm]: This check cannot possibly fail for not one, but two different reasons

File: code/modules/credits/episode_name.dm [hunk 2]
```diff
@@ -56,7 +56,7 @@
...
-	"THE CURIOUS CASE OF [uppr_name]", "ONE HELL OF A PARTY", "FOR YOUR CONSIDERATION", "PRESS YOUR LUCK", "A STATION CALLED [uppr_name]", "CRIME AND PUNISHMENT", "MY DINNER WITH [uppr_name]", "UNFINISHED BUSINESS", "THE ONLY STATION THAT'S NOT ON FIRE (YET)", "SOMEONE'S GOTTA DO IT", "THE [uppr_name] MIX-UP", "PILOT", "PROLOGUE", "FINALE", "UNTITLED", "THE END")]")
+	"THE CURIOUS CASE OF [uppr_name]", "ONE HELL OF A PARTY", "FOR YOUR CONSIDERATION", "PRESS YOUR LUCK", "A STATION CALLED [uppr_name]", "CRIME AND PUNISHMENT", "MY DINNER WITH [uppr_name]", "UNFINISHED BUSINESS", "THE ONLY STATION THAT'S NOT ON FIRE (YET)", "SOMEONE'S GOTTA DO IT", "THE [uppr_name] MIX-UP", "TRULY THEY WERE, \A [uppr_name]", "PILOT", "PROLOGUE", "FINALE", "UNTITLED", "THE END")]")
```
Comment (pedr69) [inline code/modules/credits/episode_name.dm:59]: AIEEEE UNATOMIC AAAHHHH NOOOOOOOOOOOO

Comment (SECBATON-GRIFFON) [reply code/modules/credits/episode_name.dm:59]: i dunno, it's common enough to add lines to the dream flufftext in unrelated PRs, so why not here?

File: code/game/objects/items.dm (hunk 1 above)
Comment (SECBATON-GRIFFON) [reply code/game/objects/items.dm]: fixed

**Comment by SECBATON-GRIFFON** [comment]
> > tamper a turf
> > throw nuke disk
> > ????
> > shit bricks
> 
> cerebral PR

not anymore, has to be someone else throwing it now

**Comment by Exxion** [comment]
Why are you changing the API of `throw_impact()` instead of just checking `isturf(impacted_atom)`

EDIT: Oh, I see why you're changing the API, but this seems like a very small feature to do that for

**Comment by SECBATON-GRIFFON** [comment]
> Why are you changing the API of `throw_impact()` instead of just checking `isturf(impacted_atom)`
> 
> EDIT: Oh, I see why you're changing the API, but this seems like a very small feature to do that for

it's to stop weird bad things happening with qdel()

**Comment by Exxion** [comment]
Yeah, I saw that. This still seems like a very small feature to require a change to every implementation of `throw_impact()`. And why does it stop execution of child procs by *default*? Wouldn't it make more sense to do that when it returns true, rather than false? Preferably, neither change would be made, but defaulting to allowing child calls would be better.
Also, yet again, it should check `isturf(impacted_atom)` so that this happens on the *last* call to `throw_impact()` rather than the *first*

**Comment by SECBATON-GRIFFON** [comment]
> Also, yet again, it should check `isturf(impacted_atom)` so that this happens on the _last_ call to `throw_impact()` rather than the _first_

fixed it to be this, works much more reliably now.

**Comment by Exxion** [comment]
I still don't like the changes to `throw_impact()` in general and especially not that when the parent returns the *default* (null), the child doesn't get to run its own code

**Comment by SECBATON-GRIFFON** [comment]
changed the children object running code to be if it returns the default instead of not

File: code/game/objects/items/weapons/boomerang.dm [hunk 3]
```diff
@@ -90,4 +90,4 @@
 				throwing = 0
 				return
 
-	return ..()
+	return !..()
```
Comment (Exxion) [inline code/game/objects/items/weapons/boomerang.dm]: Was this on purpose?

Comment (SECBATON-GRIFFON) [reply code/game/objects/items/weapons/boomerang.dm]: probably not, will fix

**Pull request merged**